File: packages/compose/src/main/kotlin/ai/mobileagentos/compose/agent/MobileAgentRuntime.kt
package ai.mobileagentos.compose.agent

import ai.mobileagentos.compose.feedback.FeedbackLoop
import ai.mobileagentos.compose.feedback.FeedbackEventType
import ai.mobileagentos.compose.readiness.MobileAIReadiness
import ai.mobileagentos.compose.readiness.ReadinessInput

class MobileAgentRuntime(private val config: AgentConfig) {
    private val session: AgentSession = AgentSession.create(config.userId)
    private val feedback: FeedbackLoop = FeedbackLoop(session.id)
    private val readiness: MobileAIReadiness = MobileAIReadiness()
    
    suspend fun run(input: AgentInput): AgentOutput {
        // 1. Input received
        val event1 = AgentEvent.create(
            AgentEventType.INPUT_RECEIVED,
            session.id,
            input
        )
        session.events.add(event1)
        
        var processedInput = input.input
        var contextSnapshot: Any? = null
        var memories: List<Any>? = null
        var privacyResult: ai.mobileagentos.compose.privacy.PrivacyResult? = null
        
        // 2. Context snapshot
        if (config.context != null) {
            contextSnapshot = config.context.snapshot(config.userId)
            val event2 = AgentEvent.create(
                AgentEventType.CONTEXT_SNAPSHOT_CREATED,
                session.id,
                contextSnapshot
            )
            session.events.add(event2)
        }
        
        // 3. Memory retrieval
        if (config.memory != null) {
            val retrievedMemories = config.memory.retrieveMemories(
                config.userId,
                input.input,
                5
            )
            memories = retrievedMemories
            val event3 = AgentEvent.create(
                AgentEventType.MEMORY_RETRIEVED,
                session.id,
                mapOf("count" to retrievedMemories.size)
            )
            session.events.add(event3)
        }
        
        // 4. Privacy check
        if (config.privacy != null) {
            privacyResult = config.privacy.redact(input.input)
            processedInput = privacyResult.redactedText
            val event4 = AgentEvent.create(
                AgentEventType.PRIVACY_CHECKED,
                session.id,
                mapOf(
                    "safe" to privacyResult.safe,
                    "redactions" to privacyResult.redactions.size
                )
            )
            session.events.add(event4)
            
            if (!privacyResult.safe) {
                feedback.recordEvent(
                    FeedbackEventType.PRIVACY_REDACTED,
                    mapOf("redactions" to privacyResult.redactions.size)
                )
            }
        }
        
        // 5. Agent response (deterministic for v0.2)
        val response = "Processed: $processedInput"
        val event5 = AgentEvent.create(
            AgentEventType.AGENT_RESPONSE_GENERATED,
            session.id,
            mapOf("response" to response)
        )
        session.events.add(event5)
        
        // 6. Actions suggested (empty for v0.2)
        val actions = emptyList<ai.mobileagentos.compose.actions.AgentAction>()
        val event6 = AgentEvent.create(
            AgentEventType.ACTIONS_SUGGESTED,
            session.id,
            mapOf("actionCount" to actions.size)
        )
        session.events.add(event6)
        
        // 7. Readiness calculation
        val readinessScore = readiness.calculateReadiness(
            ReadinessInput(
                contextEnabled = config.context != null,
                memoryEnabled = config.memory != null,
                privacyEnabled = config.privacy != null,
                actionsEnabled = config.actions != null,
                offlineSupport = "partial",
                uxContinuity = "good",
                appStoreReadiness = "review_needed"
            )
        )
        val event7 = AgentEvent.create(
            AgentEventType.READINESS_CALCULATED,
            session.id,
            mapOf("score" to readinessScore.overall)
        )
        session.events.add(event7)
        
        // 8. Session completed
        val event8 = AgentEvent.create(
            AgentEventType.SESSION_COMPLETED,
            session.id,
            null
        )
        session.events.add(event8)
        
        feedback.recordEvent(FeedbackEventType.SESSION_COMPLETED)
        
        return AgentOutput(
            response = response,
            actions = actions,
            readinessScore = readinessScore.overall,
            session = session,
            events = session.events.toList(),
            privacyResult = privacyResult,
            retrievedMemories = memories as? List<ai.mobileagentos.compose.memory.MemoryItem>
        )
    }
    
    fun getSession(): AgentSession = session
    
    fun getFeedback(): FeedbackLoop = feedback
    
    fun getConfig(): AgentConfig = config
}

File: packages/swift/Sources/MobileAgentOS/Agent/MobileAgentRuntime.swift
import Foundation

/// Main runtime that orchestrates the agent pipeline
@MainActor
public final class MobileAgentRuntime {
    private let config: AgentConfig
    private let contextEngine: ContextEngine
    private let memoryEngine: MemoryEngine
    private let privacyGuard: PrivacyGuard
    private let actionRouter: ActionRouter
    private let feedbackLoop: FeedbackLoop
    private let readinessCalculator: MobileAIReadiness
    
    private var currentSession: AgentSession?
    
    public init(config: AgentConfig) {
        self.config = config
        self.contextEngine = ContextEngine()
        self.memoryEngine = MemoryEngine(store: InMemoryStore())
        self.privacyGuard = PrivacyGuard()
        self.actionRouter = ActionRouter()
        self.feedbackLoop = FeedbackLoop()
        self.readinessCalculator = MobileAIReadiness()
    }
    
    public func run(_ input: AgentInput) async -> AgentOutput {
        var session = AgentSession(userId: config.userId)
        currentSession = session
        
        // Event 1: Input received
        let inputEvent = AgentEvent(type: .inputReceived, metadata: ["text": input.text])
        session.addEvent(inputEvent)
        feedbackLoop.trackEvent(inputEvent)
        
        // Event 2: Context snapshot
        let context = config.enableContext ? contextEngine.captureSnapshot() : nil
        let contextEvent = AgentEvent(type: .contextSnapshotCreated, metadata: [
            "platform": context?.platform ?? "unknown"
        ])
        session.addEvent(contextEvent)
        feedbackLoop.trackEvent(contextEvent)
        
        // Event 3: Memory retrieval
        var memories: [MemoryItem] = []
        if config.enableMemory {
            memories = await memoryEngine.retrieve(query: input.text, limit: 5)
        }
        let memoryEvent = AgentEvent(type: .memoryRetrieved, metadata: [
            "count": "\(memories.count)"
        ])
        session.addEvent(memoryEvent)
        feedbackLoop.trackEvent(memoryEvent)
        
        // Event 4: Privacy check
        var finalText = input.text
        var privacyResult: PrivacyResult?
        if config.enablePrivacy {
            privacyResult = privacyGuard.check(text: input.text)
            finalText = privacyResult?.redactedText ?? input.text
        }
        let privacyEvent = AgentEvent(type: .privacyChecked, metadata: [
            "redactions": "\(privacyResult?.redactions.count ?? 0)",
            "isSafe": "\(privacyResult?.isSafe ?? true)"
        ])
        session.addEvent(privacyEvent)
        feedbackLoop.trackEvent(privacyEvent)
        
        // Event 5: Generate response (deterministic for v0.3)
        let response = "Processed: \(finalText)"
        let responseEvent = AgentEvent(type: .agentResponseGenerated, metadata: [
            "length": "\(response.count)"
        ])
        session.addEvent(responseEvent)
        feedbackLoop.trackEvent(responseEvent)
        
        // Event 6: Actions suggested
        var actions: [AgentAction] = []
        if config.enableActions {
            actions = await actionRouter.suggestActions(input: finalText, context: context)
        }
        let actionsEvent = AgentEvent(type: .actionsSuggested, metadata: [
            "count": "\(actions.count)"
        ])
        session.addEvent(actionsEvent)
        feedbackLoop.trackEvent(actionsEvent)
        
        // Event 7: Calculate readiness
        let readinessScore = readinessCalculator.calculate(
            hasContext: context != nil,
            memoryCount: memories.count,
            privacyChecked: privacyResult != nil,
            actionCount: actions.count
        )
        let readinessEvent = AgentEvent(type: .readinessCalculated, metadata: [
            "score": "\(readinessScore)"
        ])
        session.addEvent(readinessEvent)
        feedbackLoop.trackEvent(readinessEvent)
        
        // Event 8: Session completed
        session.complete()
        let completedEvent = AgentEvent(type: .sessionCompleted, metadata: [:])
        session.addEvent(completedEvent)
        feedbackLoop.trackEvent(completedEvent)
        
        currentSession = session
        
        return AgentOutput(
            response: response,
            actions: actions,
            readinessScore: readinessScore,
            events: session.events,
            sessionId: session.id
        )
    }
    
    public func getSession() -> AgentSession? {
        return currentSession
    }
}